algorithms/kernel/pca/transform/oneapi/pca_transform_dense_default_batch_oneapi_impl.i

```diff
@@ -0,0 +1,243 @@
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  Common functions of PCA transformation on GPU
+//--
+*/
+
+#ifndef __PCA_TRANSFORM_DENSE_DEFAULT_BATCH_ONEAPI_IMPL_I__
+#define __PCA_TRANSFORM_DENSE_DEFAULT_BATCH_ONEAPI_IMPL_I__
+
+#include "cl_kernels/pca_transform_cl_kernels.cl"
+
+using namespace daal::services;
+using namespace daal::internal;
+using namespace daal::oneapi::internal;
+using namespace daal::data_management;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace pca
+{
+namespace transform
+{
+namespace oneapi
+{
+namespace internal
+{
+
+using namespace daal::oneapi::internal;
+
+template<typename algorithmFPType, transform::Method method>
+void TransformKernelOneAPI<algorithmFPType, method>::computeTransformedBlock
+        (DAAL_INT numRows, DAAL_INT numFeatures, DAAL_INT numComponents,
+         daal::oneapi::internal::UniversalBuffer & dataBlock,
+         daal::oneapi::internal::UniversalBuffer & eigenvectors,
+         services::Buffer<algorithmFPType> resultBlock)
+{
+    /* GEMM parameters */
+    algorithmFPType one  = 1.0;
+    algorithmFPType zero = 0.0;
+
+    BlasGpu<algorithmFPType>::xgemm(math::Layout::ColMajor, math::Transpose::Trans, math::Transpose::NoTrans, numComponents, numRows, numFeatures,
+        1.0, eigenvectors, numFeatures, 0, dataBlock, numFeatures, 0, 0.0, resultBlock, numComponents, 0);
+
+} /* void TransformKernelOneAPI<algorithmFPType, defaultDense>::computeTransformedBlock */
+
+
+template<typename algorithmFPType, transform::Method method>
+services::Status TransformKernelOneAPI<algorithmFPType, method>::computeInvSigmas(ExecutionContextIface& context,
+                                                                      const KernelPtr & computeInvSigmasKernel,
+                                                                      NumericTable* variances,
+                                                                      const services::Buffer<algorithmFPType> & invSigmas,
+                                                                      const size_t numFeatures)
+{
+    services::Status status;
+
+    if (variances != nullptr)
+    {
+        BlockDescriptor<algorithmFPType> varBlock;
+        variances->getBlockOfRows(0, numFeatures, readOnly, varBlock);
+
+        KernelArguments args(2);
+        args.set(0, varBlock.getBuffer(), AccessModeIds::read);
+        args.set(1, invSigmas, AccessModeIds::write);
+        KernelRange range(numFeatures);
+        context.run(range, computeInvSigmasKernel, args, &status);
+        variances->releaseBlockOfRows(varBlock);
+    }
+    return status;
+}
+
+template<typename algorithmFPType, transform::Method method>
+services::Status TransformKernelOneAPI<algorithmFPType, method>::compute(NumericTable& data, NumericTable& eigenvectors,
```

Comment:
This method is too long. I would try to decompose it into more small functions.

Reply:
You can decompose it more. Extract method when you build kernels. Extract methods for array initializations. Make `compute()` more abstract.